src/render/draw_symbol.js

```diff
@@ -140,36 +145,63 @@ function setSymbolDrawState(program, painter, isText, isSDF, rotateWithMap, pitc
+    const isZoomConstant = layer.isLayoutValueZoomConstant(sizeProp);
+    gl.uniform1i(program.u_is_size_zoom_constant, isZoomConstant ? 1 : 0);
+    gl.uniform1i(program.u_is_size_feature_constant, isFeatureConstant ? 1 : 0);
+
+    if (!isZoomConstant && !isFeatureConstant) {
+        // composite function
+        const lower = sizeStopZoomLevels[0];
+        const upper = sizeStopZoomLevels[1];
+        const t = (tr.zoom - lower) / (upper - lower);
+        gl.uniform1f(program.u_size_t, util.clamp(t, 0, 1));
+    } else if (isFeatureConstant && !isZoomConstant) {
+        // Even though we could get the exact value of the camera function at
+        // z = tr.zoom, we intentionally do not: instead, we interpolate between
+        // the camera function values at z = tileZoom and z = tileZoom + 1
+        // to be consistent with this restriction on composite functions
+        const lowerZoom = sizeStopZoomLevels[0];
+        const upperZoom = sizeStopZoomLevels[1];
+        const lowerValue = layer.getLayoutValue(sizeProp, {zoom: lowerZoom});
+        const upperValue = layer.getLayoutValue(sizeProp, {zoom: upperZoom});
+        const t = util.clamp((tr.zoom - lowerZoom) / (upperZoom - lowerZoom), 0, 1);
+        const size = lowerValue + (upperValue - lowerValue) * t;
+        gl.uniform1f(program.u_size, size);
+        gl.uniform1f(program.u_layout_size, layoutSize);
+    } else if (isFeatureConstant && isZoomConstant) {
+        const size = layer.getLayoutValue(sizeProp, {zoom: tr.zoom});
+        gl.uniform1f(program.u_size, size);
+    }
 }
 
-function drawTileSymbols(program, painter, layer, tile, buffers, isText, isSDF,
-        pitchWithMap, size) {
+function drawTileSymbols(program, programConfiguration, painter, layer, tile, buffers, isText, isSDF, pitchWithMap) {
 
     const gl = painter.gl;
     const tr = painter.transform;
 
-    const fontScale = size / (isText ? 24 : 1);
-
     if (pitchWithMap) {
-        const s = pixelsToTileUnits(tile, fontScale, tr.zoom);
+        const s = pixelsToTileUnits(tile, 1, tr.zoom);
         gl.uniform2f(program.u_extrude_scale, s, s);
     } else {
-        const s = tr.cameraToCenterDistance * fontScale;
-        gl.uniform2f(program.u_extrude_scale, tr.pixelsToGLUnits[0] * s, tr.pixelsToGLUnits[1] * s);
+        const s = tr.cameraToCenterDistance;
+        gl.uniform2f(program.u_extrude_scale,
+            tr.pixelsToGLUnits[0] * s,
+            tr.pixelsToGLUnits[1] * s);
     }
 
     if (isSDF) {
         const haloWidthProperty = `${isText ? 'text' : 'icon'}-halo-width`;
         const hasHalo = !layer.isPaintValueFeatureConstant(haloWidthProperty) || layer.paint[haloWidthProperty];
-        const gammaScale = fontScale * (pitchWithMap ? Math.cos(tr._pitch) : 1) * tr.cameraToCenterDistance;
-        gl.uniform1f(program.u_font_scale, fontScale);
+        const gammaScale = (isText ? 1 / 24 : 1) * (pitchWithMap ? Math.cos(tr._pitch) : 1) * tr.cameraToCenterDistance;
```

Comment:
Good question.  I think I made this change before I'd introduced `u_is_text` and the [fontScale calculation](https://github.com/mapbox/mapbox-gl-js/pull/4455/files#diff-3b132f79f9e50d40cccc486b7eb73fa4R28) into the shader.  Now that those are in place, the `(isText ? 1 / 24 : 1)` can be removed here in favor of multiplying `u_gamma_scale` by `fontScale` instead of by `v_size`.